shards/aggregate.go

```diff
@@ -0,0 +1,137 @@
+package shards
+
+import (
+	"sync"
+	"time"
+
+	"github.com/sourcegraph/zoekt"
+	"github.com/sourcegraph/zoekt/stream"
+)
+
+// collectSender is a sender that will aggregate results. Once sending is
+// done, you call Done to returned the aggregated result which are ranked.
+type collectSender struct {
+	aggregate          *zoekt.SearchResult
+	maxDocDisplayCount int
+}
+
+func newCollectSender(opts *zoekt.SearchOptions) *collectSender {
+	return &collectSender{
+		aggregate: &zoekt.SearchResult{
+			RepoURLs:      map[string]string{},
+			LineFragments: map[string]string{},
+		},
+		maxDocDisplayCount: opts.MaxDocDisplayCount,
+	}
+}
+
+func (c *collectSender) Send(r *zoekt.SearchResult) {
+	c.aggregate.Stats.Add(r.Stats)
+
+	if len(r.Files) > 0 {
+		c.aggregate.Files = append(c.aggregate.Files, r.Files...)
+
+		for k, v := range r.RepoURLs {
+			c.aggregate.RepoURLs[k] = v
+		}
+		for k, v := range r.LineFragments {
+			c.aggregate.LineFragments[k] = v
+		}
+	}
+
+	// The priority of our aggregate is the largest priority we collect.
+	if c.aggregate.Priority < r.Priority {
+		c.aggregate.Priority = r.Priority
+	}
+
+	// We receive monotonically decreasing values, so we update on every call.
+	//
+	// TODO we need to think about only updating this if smaller. But then we
+	// need to handle no events not sending invalid Progress.
+	c.aggregate.MaxPendingPriority = r.MaxPendingPriority
```

Comment:
we calculate max.pending() for every result coming from streamSearch. We already get a monotonically decreasing sequence of MaxPendingPriority, don't we?

Reply:
yeah, I was wondering if we should just ensure that in this layer by checking. But I like just relying on that invariant and then documenting it depends on it?